agent/taskresource/asmauth/asmauth_test.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/aws/amazon-ecs-agent/agent/credentials/mocks"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	executionCredentialsID = "exec-creds-id"
+	region                 = "us-west-2"
+	secretID               = "meaning-of-life"
+	username               = "irene"
+	password               = "sher"
+	taskARN                = "task1"
+)
+
+var (
+	asmAuthDataVal       string
+	requiredASMResources []*apicontainer.ASMAuthData
+)
+
+func init() {
+	asmAuthDataBytes, _ := json.Marshal(&asm.AuthDataValue{
+		Username: aws.String(username),
+		Password: aws.String(password),
+	})
+	asmAuthDataVal = string(asmAuthDataBytes)
+	requiredASMResources = []*apicontainer.ASMAuthData{
+		{
+			CredentialsParameter: secretID,
+			Region:               region,
+		},
+	}
+}
+
+func TestCreateAndGet(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	credentialsManager := mock_credentials.NewMockManager(ctrl)
+	asmClientCreator := mock_factory.NewMockClientCreator(ctrl)
+	mockASMClient := mock_secretsmanageriface.NewMockSecretsManagerAPI(ctrl)
+
+	iamRoleCreds := credentials.IAMRoleCredentials{}
+	creds := credentials.TaskIAMRoleCredentials{
+		IAMRoleCredentials: iamRoleCreds,
+	}
+	asmSecretValue := &secretsmanager.GetSecretValueOutput{
+		SecretString: aws.String(asmAuthDataVal),
+	}
+	gomock.InOrder(
+		credentialsManager.EXPECT().GetTaskCredentials(executionCredentialsID).Return(creds, true),
+		asmClientCreator.EXPECT().NewASMClient(region, iamRoleCreds).Return(mockASMClient),
+		mockASMClient.EXPECT().GetSecretValue(gomock.Any()).Do(func(in *secretsmanager.GetSecretValueInput) {
+			assert.Equal(t, aws.StringValue(in.SecretId), secretID)
+		}).Return(asmSecretValue, nil),
+	)
+	asmRes := &ASMAuthResource{
+		executionCredentialsID: executionCredentialsID,
+		requiredASMResources:   requiredASMResources,
+		credentialsManager:     credentialsManager,
+		asmClientCreator:       asmClientCreator,
+	}
+	require.NoError(t, asmRes.Create())
+	dac, ok := asmRes.GetASMDockerAuthConfig(secretID)
+	require.True(t, ok)
+	assert.Equal(t, dac.Username, username)
+	assert.Equal(t, dac.Password, password)
+}
+
+func TestMarshalUnmarshalJSON(t *testing.T) {
+	asmResIn := &ASMAuthResource{
+		taskARN:                taskARN,
+		executionCredentialsID: executionCredentialsID,
+		createdAt:              time.Now(),
+		knownStatusUnsafe:      taskresource.ResourceCreated,
+		desiredStatusUnsafe:    taskresource.ResourceCreated,
+		requiredASMResources:   requiredASMResources,
+	}
+
+	bytes, err := json.Marshal(asmResIn)
+	require.NoError(t, err)
+
+	asmResOut := &ASMAuthResource{}
+	err = json.Unmarshal(bytes, asmResOut)
+	require.NoError(t, err)
+	assert.Equal(t, asmResIn.taskARN, asmResOut.taskARN)
```

Comment:
assert does deep equal so assert.Equal(t,asmResIn,asmResOut) should work for you (I think?)

Reply:
eh why not check field by field?

Reply:
Its fine as is -- its just more concise the other way.

```
/// this... 
assert.Equal(t, expectedStruct, actualStruct)

/// instead of ...
assert.Equal(t, expectedStruct.field1, actualStruct.field1)
assert.Equal(t, expectedStruct.field2, actualStruct.field2)
assert.Equal(t, expectedStruct.field3, actualStruct.field3)
// ...
```